core-bundle/src/EventListener/DataContainer/RecordPreviewListener.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\EventListener\DataContainer;
+
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\CoreBundle\ServiceAnnotation\Hook;
+use Contao\DataContainer;
+use Contao\System;
+use Doctrine\DBAL\Connection;
+
+/**
+ * @internal
+ */
+class RecordPreviewListener
+{
+    private ContaoFramework $framework;
+
+    private Connection $connection;
+
+    public function __construct(ContaoFramework $framework, Connection $connection)
+    {
+        $this->framework = $framework;
+        $this->connection = $connection;
+    }
+
+    /**
+     * @Hook("loadDataContainer")
+     */
+    public function registerDeleteCallbacks(string $table): void
+    {
+        if ($GLOBALS['TL_DCA'][$table]['config']['notDeletable'] ?? false) {
+            return;
+        }
+
+        if (!is_array($GLOBALS['TL_DCA'][$table]['config']['ondelete_callback'])) {
+            $GLOBALS['TL_DCA'][$table]['config']['ondelete_callback'] = [];
+        }
+
+        $GLOBALS['TL_DCA'][$table]['config']['ondelete_callback'][] = [
+            'contao.listener.data_container.record_preview', 'storePrerenderedRecordPreview'
+        ];
+    }
+
+    public function storePrerenderedRecordPreview(DataContainer $dc, $undoId): void
+    {
+        try {
+            $row = $this->connection
+                ->prepare('SELECT * FROM ' . $this->connection->quoteIdentifier($dc->table) . ' WHERE id = ?')
+                ->executeQuery([$dc->id])
+                ->fetchAssociative();
+            $preview = $this->renderPreview($dc, $row);
+        } catch (\Exception $exception) {
+            $preview = '';
+        }
+
+        $this->connection->update('tl_undo', ['preview' => $preview], ['id' => $undoId]);
+    }
+
+    private function renderPreview(DataContainer $dc, array $row): string
+    {
+        $system = $this->framework->getAdapter(System::class);
+
+        if (\is_array($GLOBALS['TL_DCA'][$dc->table]['list']['sorting']['child_record_callback'] ?? null)) {
+            $class = $GLOBALS['TL_DCA'][$dc->table]['list']['sorting']['child_record_callback'][0];
+            $method = $GLOBALS['TL_DCA'][$dc->table]['list']['sorting']['child_record_callback'][1];
+
+            $instance = $system->importStatic($class);
+            $preview = $instance->{$method}($row);
+        } elseif (\is_callable($GLOBALS['TL_DCA'][$dc->table]['list']['sorting']['child_record_callback'] ?? null)) {
+            $preview = $GLOBALS['TL_DCA'][$dc->table]['list']['sorting']['child_record_callback']($row);
```

Comment:
Technically you should check for the DC mode here, as the child record callback is only executed for the parent modes and is otherwise ignored.